### PR TITLE
[evm] Get explorer, node and gasTracker information from coin config

### PR DESCRIPTION
**Pull request by lvndry**
### 📝 Description

Use coin config to get explorerViews information

Default value:

Screenshot 2024-03-08 at 17 27 18

After changing the node uri in Firebase

Screenshot 2024-03-08 at 17 32 50

### ❓ Context

- **JIRA or GitHub link**: https://ledgerhq.atlassian.net/browse/LIVE-11477

### ✅ Checklist

Pull Requests must pass the CI and be code reviewed. Set as Draft if the PR is not ready.

- [ ] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - ...

---

### 🧐 Checklist for the PR Reviewers

- [ ] **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- [ ] **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- [ ] **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- [ ] **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- [ ] **Any new dependencies** have been justified and documented.
- [ ] **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: libs/coin-evm/src/api/explorer/types.ts [hunk 1]
```diff
@@ -16,7 +17,7 @@ export type ExplorerApi = {
   }>;
 };
 
-type ExplorerConfig = EthereumLikeInfo["explorer"];
+type ExplorerConfig = EvmConfig["explorer"];
```
Comment (sprohaszka-ledger) [inline libs/coin-evm/src/api/explorer/types.ts:20]: Not sure you need to change this. What we want is to be sure that the `ExplorerConfig` is the kind describe in `EthereumLikeInfo`

File: libs/coin-evm/src/api/gasTracker/types.ts [hunk 2]
```diff
@@ -13,7 +14,7 @@ export type GasTrackerApi = {
   }) => Promise<GasOptions>;
 };
 
-type GasTrackerConfig = EthereumLikeInfo["gasTracker"];
+type GasTrackerConfig = EvmConfigInfo["gasTracker"];
```
Comment (sprohaszka-ledger) [inline libs/coin-evm/src/api/gasTracker/types.ts:17]: Same comment about where should be the source of truth about the type definition (maybe more in `EthereumLikeInfo`)

File: libs/coin-evm/src/api/node/ledger.ts [hunk 3]
```diff
@@ -223,7 +231,6 @@ export const getFeeData: NodeApi["getFeeData"] = async (currency, transaction) =
       ...currency,
       ethereumLikeInfo: {
         ...currency.ethereumLikeInfo!,
-        gasTracker: { type: "ledger", explorerId: node.explorerId },
```
Comment (sprohaszka-ledger) [inline libs/coin-evm/src/api/node/ledger.ts:226]: Why removing it?

File: libs/coin-evm/src/api/node/types.ts [hunk 4]
```diff
@@ -34,7 +35,7 @@ export type NodeApi = {
   ) => Promise<BigNumber>;
 };
 
-type NodeConfig = EthereumLikeInfo["node"];
+type NodeConfig = EvmConfigInfo["node"];
```
Comment (sprohaszka-ledger) [inline libs/coin-evm/src/api/node/types.ts:38]: (same comment about type source of truth)

File: libs/coin-evm/src/logic.ts [hunk 5]
```diff
@@ -199,7 +200,9 @@ export const getSyncHash = (
     .map(token => token.id + token.contractAddress + token.name + token.ticker)
     .join("");
   const isNftSupported = isNFTActive(currency);
-  const { node = {}, explorer = {} } = currency.ethereumLikeInfo || {};
+
+  const config = getCoinConfig(currency).info;
+  const { node = {}, explorer = {} } = config;
```
Comment (sprohaszka-ledger) [inline libs/coin-evm/src/logic.ts:205]: Do you think we still need this default value? Otherwise we can clean it at the same time.

File: libs/ledger-live-common/src/config/types.ts [hunk 6]
```diff
@@ -1 +1 @@
-export type CurrencyConfigCommon = Record<string, unknown>;
+export type CurrencyConfigCommon<Config = Record<string, unknown>> = Config;
```
Comment (sprohaszka-ledger) [inline libs/ledger-live-common/src/config/types.ts:1]: I am not fan of the use of generics for composition :/

File: libs/coin-framework/src/config.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export type CommonCurrencyConfig = {
+  status: {
+    type: "active" | "under_maintenance";
+  };
+};
+
+export type CurrencyConfig<Config = Record<string, unknown>> = CommonCurrencyConfig & Config;
```
Comment (sprohaszka-ledger) [inline libs/coin-framework/src/config.ts:7]: Same comment about using generics for composition

File: libs/ledger-live-common/src/explorers.ts [hunk 8]
```diff
@@ -1,29 +1,37 @@
...
+};
+
 export const getAddressExplorer = (
-  explorerView: ExplorerView | null | undefined,
+  explorerView: ExplorerView | undefined,
   address: string,
-): string | null | undefined =>
-  explorerView && explorerView.address && explorerView.address.replace("$address", address);
+): string | undefined => {
+  return explorerView?.address?.replace("$address", address);
```
Comment (sprohaszka-ledger) [inline libs/ledger-live-common/src/explorers.ts:19]: 👍

File: libs/ledger-live-common/src/families/evm/setup.ts [hunk 9]
```diff
@@ -15,12 +15,22 @@ import {
 import { Resolver } from "../../hw/getAddress/types";
 import Transport from "@ledgerhq/hw-transport";
 import { Bridge } from "@ledgerhq/types-live";
+import { getCurrencyConfiguration } from "../../config";
+import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
+import { EvmConfigInfo } from "@ledgerhq/coin-evm/lib/config";
```
Comment (sprohaszka-ledger) [inline libs/ledger-live-common/src/families/evm/setup.ts:20]: I think the import should work without the `lib` path.

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts [hunk 10]
```diff
@@ -265,9 +265,6 @@ export const cryptocurrenciesById: Record<CryptoCurrencyId, CryptoCurrency> = {
     ],
     ethereumLikeInfo: {
       chainId: 43114,
-      node: { type: "ledger", explorerId: "avax" },
```
Comment (sprohaszka-ledger) [inline libs/ledgerjs/packages/cryptoassets/src/currencies.ts:268]: Just to be sure about the behaviour: we don't want to keep them as a default value if CoinConfig is not set or loaded?

**Review by sprohaszka-ledger — COMMENTED** [review]
I wonder where should be the source of truth about the definition EthereumLikeInfo.

File: libs/coin-evm/src/api/explorer/types.ts (hunk 1 above)
Comment (lvndry) [reply libs/coin-evm/src/api/explorer/types.ts:20]: Actually the EthereumLikeInfo object only has `{ chainId: number}` so `explorer` and `node` are only information that can be found in the config

File: libs/coin-evm/src/api/node/ledger.ts (hunk 3 above)
Comment (lvndry) [reply libs/coin-evm/src/api/node/ledger.ts:226]: We get the gasTracker information directly in `getGasOptions`

File: libs/coin-evm/src/logic.ts (hunk 5 above)
Comment (lvndry) [reply libs/coin-evm/src/logic.ts:205]: Yes the default values are still needed because node and explorer are not in the config for every evm currency. However we use the objects line 219 to create the hash

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts (hunk 10 above)
Comment (lvndry) [reply libs/ledgerjs/packages/cryptoassets/src/currencies.ts:268]: There's always default values set here:
https://github.com/LedgerHQ/ledger-live/blob/feat/live-1147-evm-coin-config-explorer/libs/ledger-live-common/src/families/evm/config.ts

File: libs/coin-evm/src/__tests__/fixtures/synchronization.fixtures.ts [hunk 11]
```diff
@@ -20,17 +21,24 @@ export const currency: CryptoCurrency = Object.freeze({
...
-      uri: "https://my-rpc.com",
-    },
-    explorer: {
-      type: "etherscan" as const,
-      uri: "https://api.com",
-    },
   },
 });
 
+setCoinConfig((): any => {
```
Comment (hedi-edelbloute) [inline libs/coin-evm/src/__tests__/fixtures/synchronization.fixtures.ts:27]: I'm scared this edits the coinConfig globally, Ideally we want to mock the returned value of getConfig call we can do this together :)

another dirty solution is to store in beforeAll and reapply in afterAll but not clean

To apply on every tests

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts [hunk 12]
```diff
@@ -3716,11 +3689,6 @@ export const cryptocurrenciesById: Record<CryptoCurrencyId, CryptoCurrency> = {
     units: ethereumUnits("ETH", "ETH"),
     ethereumLikeInfo: {
       chainId: 42161,
```
Comment (hedi-edelbloute) [inline libs/ledgerjs/packages/cryptoassets/src/currencies.ts:3691]: This is the latest element in currencies.ts ethereumLikeInfo :thinking: We could add it in coinconfig but this won't be updated and i'm not sure this makes sense... WDYT ?

I think long term a lot more will be pulled remotely soooo

**Comment by hedi-edelbloute** [comment]
Missing a changeset btw !

File: libs/ledgerjs/packages/cryptoassets/src/currencies.ts (hunk 12 above)
Comment (lvndry) [reply libs/ledgerjs/packages/cryptoassets/src/currencies.ts:3691]: I don't think it's relevant to have the chainId as a "config", it's an attribute of a currency

File: libs/coin-evm/src/__tests__/fixtures/synchronization.fixtures.ts (hunk 11 above)
Comment (lvndry) [reply libs/coin-evm/src/__tests__/fixtures/synchronization.fixtures.ts:27]: In the tests we usually have a `beforeEach` that overrides this one. but it's still needed before in the file we call `getSyncHash` that requires a coinConfig set up

Comment (sprohaszka-ledger) [reply libs/coin-evm/src/__tests__/fixtures/synchronization.fixtures.ts:27]: I tend to agree with Hedi here.
If you don't want to use a mock, call a `afterAll` to erase the CoinConfig setted this way. It will avoid side effects.

**Review by hedi-edelbloute — APPROVED** [review]

File: libs/coin-evm/src/createTransaction.ts
Comment (lambertkevin) [inline libs/coin-evm/src/createTransaction.ts:1]: Unnecessary change

File: libs/coin-evm/src/api/node/ledger.ts (hunk 3 above)
Comment (lambertkevin) [reply libs/coin-evm/src/api/node/ledger.ts:226]: No actually @sprohaszka-ledger has a point, this removal is kinda problematic. 
This overriding mechanism is here to help with the separation of concerns of the 3 "backends" we can rely on with an EVM: A node, and explorer and a gas tracker.

Here, the goal was to say that Ledger "as a node" should be able to provide the exact same methods as a classic RPC node. But Ledger's backend isn't providing all features for now. In order to fake the behaviour of an eth_gasPrice/eth_feeHistory RPC request, we go get the result of the Ledger `gasTracker` (no matter what the configuration of the currency is) and return it as if it was an eth_gasPrice/eth_feeHistory RPC response.

Here if the configuration of the coin is set to use let's say Blocknative gas tracker (not implemented today, I know), using Ledger "as a node" would result in using Blocknative when requesting fees, and not just Ledger anymore.

So we need a way to enforce this ledger config when using `getGasOptions`.

**Review by lambertkevin — CHANGES_REQUESTED** [review]

**Review by gre — DISMISSED** [review]
ok for hub scope

**Review by lambertkevin — APPROVED** [review]
